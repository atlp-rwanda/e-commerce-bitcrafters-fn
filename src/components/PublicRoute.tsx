import React from "react";
import { Navigate, Outlet } from "react-router-dom";
import { useSelector, shallowEqual } from "react-redux";
import { RootState } from "../redux/store";
import { toast } from 'react-toastify';

const PublicRoute: React.FC = () => {
  const isLoggedIn = useSelector(
    (state: RootState) => state.auth.isLoggedIn,
    shallowEqual,
  );

  if (isLoggedIn) {
    return <Navigate to="/" />;
  }

  return <Outlet />;
};
export const PrivateRoute: React.FC = () => {
  const isLoggedIn = useSelector(
    (state: RootState) => state.auth.isLoggedIn,
    shallowEqual,
  );

  if (!isLoggedIn) {
    toast.error("Please Login")
    return <Navigate to="/" />;
  }

  return <Outlet />;
};

export default PublicRoute;
