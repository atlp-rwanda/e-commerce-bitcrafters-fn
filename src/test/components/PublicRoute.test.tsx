// PublicRoute.test.tsx
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { BrowserRouter } from "react-router-dom";
import PublicRoute, { PrivateRoute } from "../../components/PublicRoute";
import { toast } from 'react-toastify';

// Mock the toast module
jest.mock('react-toastify', () => ({
  toast: {
    error: jest.fn(),
  },
}));

// Create mock store
const mockStore = configureStore([]);
const initialState = {
  auth: {
    isLoggedIn: false,
  },
};

describe("Route Components", () => {
  let store: any;

  beforeEach(() => {
    store = mockStore(initialState);
  });

  it("renders PublicRoute correctly when not logged in", () => {
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <PublicRoute />
        </BrowserRouter>
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it("redirects to home page when logged in on PublicRoute", () => {
    store = mockStore({
      auth: {
        isLoggedIn: true,
      },
    });

    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <PublicRoute />
        </BrowserRouter>
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it("renders PrivateRoute correctly when logged in", () => {
    store = mockStore({
      auth: {
        isLoggedIn: true,
      },
    });

    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <PrivateRoute />
        </BrowserRouter>
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it("redirects to login page and shows toast when not logged in on PrivateRoute", () => {
    store = mockStore({
      auth: {
        isLoggedIn: false,
      },
    });

    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <PrivateRoute />
        </BrowserRouter>
      </Provider>
    );

    expect(toast.error).toHaveBeenCalledWith("Please Login");
    expect(container).toMatchSnapshot();
  });
});
