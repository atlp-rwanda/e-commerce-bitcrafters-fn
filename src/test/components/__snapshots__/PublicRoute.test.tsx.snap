// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route Components redirects to home page when logged in on PublicRoute 1`] = `<div />`;

exports[`Route Components redirects to login page and shows toast when not logged in on PrivateRoute 1`] = `<div />`;

exports[`Route Components renders PrivateRoute correctly when logged in 1`] = `<div />`;

exports[`Route Components renders PublicRoute correctly when not logged in 1`] = `<div />`;
